rotkehlchen/chain/ethereum/modules/balancer/v1/decoder.py

```diff
@@ -0,0 +1,289 @@
+                asset = event.asset.resolve_to_asset_with_symbol()
+                event.notes = f'Refunded {event.balance.amount} {asset.symbol} after depositing in Balancer V1 pool'  # noqa: E501
+
+        return decoded_events
+
+    def _check_deposits_withdrawals_v1(
+            self,
+            transaction: EvmTransaction,  # pylint: disable=unused-argument
+            decoded_events: list[HistoryBaseEntry],
+            all_logs: list[EvmTxReceiptLog],  # pylint: disable=unused-argument
+            counterparties: set[str],
+    ) -> list[HistoryBaseEntry]:
+        """
+        Check for accounting in v1 that the deposits/withdrawals events have the needed information
+        to process them during accounting.
+        """
+        if CPT_BALANCER_V1 not in counterparties:
+            return decoded_events
+
+        related_events: list[HistoryBaseEntry] = []
+        related_events_map: dict[HistoryBaseEntry, list[HistoryBaseEntry]] = {}
+        # last event is only tracked in the case of exiting a pool and contains the event
+        # sending the BPT token
+        last_event = None
+        for event in decoded_events:
+            if event.counterparty != CPT_BALANCER_V1:
+                continue
+
+            # When joining a pool first we spend the assets and then we receive the BPT token.
+            # In the case of exiting the pool first we return the BPT token and then we receive
+            # the assets.
+            # To handle properly the case of multiple events happening in the same transaction we
+            # accumulate them in the current list of events and then we empty it in the
+            # related_events map
+            if (
+                event.event_type == HistoryEventType.RECEIVE and
+                event.event_subtype == HistoryEventSubType.RECEIVE_WRAPPED
+            ):
+                if len(related_events) != 0:
+                    related_events_map[event] = related_events
+                    related_events = []
+                    last_event = None
+            elif (
+                event.event_type == HistoryEventType.SPEND and
+                event.event_subtype == HistoryEventSubType.RETURN_WRAPPED
+            ):
+                related_events = []
+                last_event = event
+            elif (
+                (
+                    event.event_type == HistoryEventType.DEPOSIT and
+                    event.event_subtype == HistoryEventSubType.DEPOSIT_ASSET
+                ) or
+                (
+                    event.event_type == HistoryEventType.RECEIVE and
+                    event.event_subtype == HistoryEventSubType.REMOVE_ASSET
+                )
+            ):
+                # In the case that we are handling a join that comes after an exit we have to
+                # check if there is a change in the direction of the assets (from spending them)
+                # to receiving them
+                if last_event is not None:
```

Comment:
here if `last_event` is None is it fine to add `event` to `related_events`? I don't really understand the logic here ... it's really hard to read

Reply:
Yes, in this case last event is not `None` when you are handling an exit and then a deposit. Adding of `event` to `related_events` is needed because you are handling a different part of the transaction, the deposit.

So the if statement makes sure that we finish any previous handling and the the append is starting the new one